### PR TITLE
Allow to select assets in the layout

### PR DESCRIPTION
**Pull request by aschempp**
Since I switched to [Symfony Encore](https://symfony.com/doc/current/frontend.html), I always need to add my CSS and JS files using custom HTML tags in the page layout.

With this PR I can select JS and CSS files from Symfony Assets in the page layout by simply registering a file that lists available options:

```
framework:
    assets:
        json_manifest_path: '%kernel.project_dir%/web/path/to/my.json'
```

### CODE REVIEW

File: core-bundle/src/Resources/contao/pages/PageRegular.php [hunk 1]
```diff
@@ -757,6 +769,18 @@ protected function createFooterScripts($objLayout)
 			}
 		}
 
+		$arrJsAssets = \StringUtil::deserialize($objLayout->jsAssets);
+
+		// External style sheets
```
Comment (xchs) [inline core-bundle/src/Resources/contao/pages/PageRegular.php]: "External **JavaScripts**", isn't it?

Comment (aschempp) [reply core-bundle/src/Resources/contao/pages/PageRegular.php]: Fixed in 1ba5f25f72a0343804fdf5dab1a823227dcb1298

**Comment by fritzmg** [comment]
:tada:

**Comment by aschempp** [comment]
@heimrichhannot I saw you have a bundle for something similar, maybe you wanna provide some feedback as well? 😊

**Comment by leofeyer** [comment]
I wonder if all the `manifest.json` related stuff should be encapsulated in a separate bundle? It seems only relevant for a small user group using Encore.

**Comment by Toflar** [comment]
It has nothing to do with Encore. It's just a web app manifest, why would you not want to support W3C standards (or drafts in this case)? https://www.w3.org/TR/appmanifest/

**Comment by leofeyer** [comment]
I do want to support it. I just don't want to clutter the core with features that are only relevant for a minority of users.

File: core-bundle/src/Resources/config/listener.yml [hunk 2]
```diff
@@ -56,6 +56,12 @@ services:
         tags:
             - { name: kernel.event_listener, event: kernel.terminate, method: onKernelTerminate }
 
+    contao.listener.data_container.layout_assets:
+        class: Contao\CoreBundle\EventListener\DataContainer\LayoutAssets
+        public: true # TODO tag the service as DCA callback
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/listener.yml]: Feel free to do this now (see https://github.com/contao/contao/pull/39). 😉

**Comment by fritzmg** [comment]
I do not understand why this is strictly tied to the `manifest.json` in the first place (according to the current label description). Does it not make sense in general to be able to add JavaScript and CSS assets from the `assets` folder to the layout? Regardless of whether you have added those assets to the `manifest.json` as well.

**Comment by aschempp** [comment]
> I wonder if all the manifest.json related stuff should be encapsulated in a separate bundle? It seems only relevant for a small user group using Encore.

It is only partially related to Encore. It's actually Symfony Assets, which we already support using an insert tag and for all contao-components. Also, I think we should promote this implementation as a new best practice for Contao.

> It has nothing to do with Encore. It's just a web app manifest, why would you not want to support W3C standards (or drafts in this case)? https://www.w3.org/TR/appmanifest/

This PR has nothing to do with *App Manifest*, this is just a json file that contains file links, which for some reason is named `manifest.json` by default in Webpack/Symfony Encore. Sorry about that confusion, I have reworded the initial description.


> Does it not make sense in general to be able to add JavaScript and CSS assets from the assets folder to the layout? Regardless of whether you have added those assets to the manifest.json as well.

I think it does, and that's pretty much on point. You can simply use any json file to register the CSS and JS files you want to have available in your page layouts. They don't need to be generated by Webpack/Encore.

File: core-bundle/src/EventListener/DataContainer/LayoutAssets.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Symfony\Component\Filesystem\Filesystem;
+
+class LayoutAssets
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/LayoutAssets.php]: The class should be called `LayoutAssetsListener`.

File: core-bundle/src/EventListener/DataContainer/LayoutAssets.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        return $this->getFilesWithExtension('.css');
+    }
+
+    public function getJsAssets(): array
+    {
+        return $this->getFilesWithExtension('.js');
+    }
+
+    private function getFilesWithExtension(string $extension)
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/LayoutAssets.php]: `: void` is missing

File: core-bundle/src/EventListener/DataContainer/LayoutAssets.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        return array_values(array_filter(
+            $this->data,
+            function ($file) use ($extension) {
+                return substr($file, strlen($extension) * -1) === $extension;
+            }
+        ));
+    }
+
+    private function load()
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainer/LayoutAssets.php]: `: void` is missing

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Comment by aschempp** [comment]
**Important:**
This has just become kinda obsolete/out of date with the new Encore update: https://symfony.com/blog/encore-0-21-0-webpack-4-support-copy-files-support-webpackencorebundle

**Comment by leofeyer** [comment]
> This has just become kinda obsolete/out of date with the new Encore update: 

Does that mean I am not supposed to merge this?

**Comment by aschempp** [comment]
Correcty, don't merge it. But let's keep it atm as a reminder to change the implementation.

FYI, Encore has introduced "entrypoints", which makes the individual selection of CSS and JS obsolete. One would only need to select an entrypoint in the layout, and Encore knows which scripts to include. They're also about to introduce async and defer support, so it would basically solve all our problems!

**Comment by koertho** [comment]
@aschempp Sorry, didn't got or overseen a notification about this thread. We still have out encore bundle which works really good: https://github.com/heimrichhannot/contao-encore-bundle

Next to add assets from layout (from version 1.0), you can also add/unset assets on page basis, which can be an additional performance increase. You really welcome to view the code and give us some feedback, if we should do pull request. But for us as company time is often to short to really think about this (and do all the preparation work).

**Comment by leofeyer** [comment]
FYI: @baumannsven is also working on an Encore integration for Contao, which allows to select entry points in the layout module in the back end.

**Comment by baumannsven** [comment]
@leofeyer Thats right. If the first release finished, i wrote a message here.

**Comment by fkaminski** [comment]
> This has just become kinda obsolete/out of date with the new Encore update: 

Question: Why?
IMHO it is a fine feature in general because of:

> ... You can simply use any json file to register the CSS and JS files you want to have available in your page layouts. ...

Furthermore the additional possibility to add/unset assets on page base level,
as implemented by @koertho /@baumannsven in their bundles, sounds good too.

**Comment by baumannsven** [comment]
@leofeyer The bundle is now released as stable version.

https://github.com/ContaoBlackForest/contao-encore-bundle

**Comment by baumannsven** [comment]
Which I noticed. Here it is mentioned that the assets are added via the page settings. This is a wrong approach. Because you can't set such things in the page in any way.

**Comment by aschempp** [comment]
I'm closing this for now, as it's not the *right* was to implement in Contao.

**Pull request closed without merging**